File: backend/roboflow_runner/roboflow_runner/flow_executor.py
"""Traverse RoboFlow JSON (same rules as js/flow-executor.js)."""

from __future__ import annotations

from typing import Callable, Dict, List, Optional

MAX_STEPS = 5000


class FlowExecutionError(Exception):
    pass


def get_next(connections: List[dict], block_id: str, port: int) -> Optional[str]:
    for c in connections:
        if c.get('sourceId') == block_id and int(c.get('sourcePort', 0)) == port:
            return c.get('targetId')
    return None


def find_start_block(blocks: List[dict]) -> Optional[dict]:
    for b in blocks:
        if b.get('type') == 'start':
            return b
    return None


def index_blocks(blocks: List[dict]) -> Dict[str, dict]:
    return {b['id']: b for b in blocks if b.get('id')}


def run_flow(
    flow: dict,
    execute_block: Callable[[str, dict], None],
    *,
    should_stop: Callable[[], bool],
    log: Callable[[str], None],
) -> None:
    blocks_list = flow.get('blocks') or []
    connections = flow.get('connections') or []
    blocks = index_blocks(blocks_list)
    start = find_start_block(blocks_list)
    if not start:
        raise FlowExecutionError('Flow has no START block')

    loop_counters: Dict[str, Dict[str, int]] = {}
    current_id: Optional[str] = start['id']
    steps = 0

    while current_id and not should_stop() and steps < MAX_STEPS:
        steps += 1
        block = blocks.get(current_id)
        if not block:
            break

        btype = block.get('type') or ''
        params = block.get('params') or {}

        if btype == 'end':
            execute_block(btype, params)
            break

        if btype == 'loop_start':
            if current_id not in loop_counters:
                raw = params.get('iterations', 3)
                try:
                    iters = int(raw)
                except (TypeError, ValueError):
                    iters = 3
                loop_counters[current_id] = {'current': 0, 'max': max(1, iters)}
            ctr = loop_counters[current_id]
            ctr['current'] += 1
            log(f'Loop {ctr["current"]}/{ctr["max"]}')
            if ctr['current'] <= ctr['max']:
                current_id = get_next(connections, current_id, 0)
            else:
                del loop_counters[current_id]
                current_id = get_next(connections, current_id, 1)
            continue

        if btype == 'loop_end':
            current_id = get_next(connections, current_id, 0)
            continue

        execute_block(btype, params)
        if should_stop():
            break
        current_id = get_next(connections, current_id, 0)

    if steps >= MAX_STEPS:
        raise FlowExecutionError('Max steps exceeded — possible infinite loop')

File: js/flow-executor.js
/**
 * RoboFlow — Flow Executor
 * Traverses the node graph and executes each block sequentially.
 * Supports Loop Start / Loop End cycles.
 */
const FlowExecutor = (() => {

  /* ── State ─────────────────────────────────────────────── */
  let running      = false;
  let paused       = false;
  let stopRequested = false;
  let loopCounters = new Map();  // blockId → { current, max }
  let currentBlockId = null;
  const MAX_STEPS = 5000;

  /* ── UI helpers ─────────────────────────────────────────── */
  function setStatus(text, cls = '') {
    const el = document.getElementById('exec-status-text');
    const log = document.getElementById('execution-log');
    if (el) el.textContent = text;
    if (log) {
      log.className = 'exec-log' + (cls ? ' ' + cls : '');
    }
    ROS.simLog(text);
  }

  function setStepBadge(text) {
    const b = document.getElementById('exec-step-badge');
    if (!b) return;
    if (text) { b.textContent = text; b.classList.remove('hidden'); }
    else       { b.classList.add('hidden'); }
  }

  function setButtonStates(isRunning) {
    const btnRun   = document.getElementById('btn-run');
    const btnPause = document.getElementById('btn-pause');
    const btnStop  = document.getElementById('btn-stop-exec');
    if (btnRun)   btnRun.disabled   = isRunning;
    if (btnPause) btnPause.disabled  = !isRunning;
    if (btnStop)  btnStop.disabled   = !isRunning;
  }

  /* ── Wait for un-pause ──────────────────────────────────── */
  function waitForUnpause() {
    return new Promise(resolve => {
      const interval = setInterval(() => {
        if (!paused || stopRequested) { clearInterval(interval); resolve(); }
      }, 100);
    });
  }

  /* ── Main run ───────────────────────────────────────────── */
  async function run() {
    if (running) return;

    // Validate: find START block
    const startBlock = FlowEngine.findBlockByType('start');
    if (!startBlock) {
      showToast('Add a START block to begin!', 'warning');
      return;
    }

    running       = true;
    paused        = false;
    stopRequested = false;
    loopCounters  = new Map();

    setButtonStates(true);
    setStatus('▶ Starting flow…', 'running');

    // Highlight start
    FlowEngine.setBlockExecuting(startBlock.id, true);

    let currentId = startBlock.id;
    let steps     = 0;

    try {
      while (currentId && !stopRequested && steps < MAX_STEPS) {
        steps++;

        if (paused) {
          setStatus('⏸ Paused', 'paused');
          await waitForUnpause();
          if (stopRequested) break;
          setStatus('▶ Resuming…', 'running');
        }

        const block = FlowEngine.getBlock(currentId);
        if (!block) break;

        const def   = FlowBlocks.get(block.type);
        if (!def)   { currentId = getNext(currentId, 0); continue; }

        // ── Highlight current block ──
        if (currentBlockId) FlowEngine.setBlockExecuting(currentBlockId, false);
        currentBlockId = currentId;
        FlowEngine.setBlockExecuting(currentId, true);
        setStepBadge(`Step ${steps}`);

        // ── END block → finish ──
        if (block.type === 'end') {
          await def.execute(block.params || {});
          break;
        }

        // ── LOOP START (two inputs: enter ← START/prior, repeat ← Loop End) ──
        if (block.type === 'loop_start') {
          if (!loopCounters.has(currentId)) {
            const max = parseInt(block.params?.iterations || def.params[0].default, 10);
            loopCounters.set(currentId, { current: 0, max });
          }
          const counter = loopCounters.get(currentId);
          counter.current++;
          setStatus(`🔄 Loop ${counter.current}/${counter.max}`, 'running');

          if (counter.current <= counter.max) {
            // Follow "body" output (port 0)
            currentId = getNext(currentId, 0);
          } else {
            // Loop exhausted → follow "exit" output (port 1)
            loopCounters.delete(currentId);
            currentId = getNext(currentId, 1);
          }
          continue;
        }

        // ── LOOP END — wire output → Loop Start “repeat” input (input port 1) ──
        if (block.type === 'loop_end') {
          currentId = getNext(currentId, 0);
          continue;
        }

        // ── Regular block ──
        setStatus(`⚙ Executing: ${def.label}`, 'running');
        try {
          await def.execute(block.params || {});
        } catch (err) {
          setStatus(`⚠ Error in "${def.label}": ${err.message}`, 'error');
          showToast(`Block error: ${err.message}`, 'error');
          break;
        }

        if (stopRequested) break;

        currentId = getNext(currentId, 0);
      }

      if (steps >= MAX_STEPS) {
        setStatus('⚠ Max steps reached — possible infinite loop', 'error');
        showToast('Execution stopped: max steps reached', 'warning');
      } else if (!stopRequested) {
        setStatus('✅ Flow complete!', 'done');
        showToast('Flow executed successfully!', 'success');
      } else {
        setStatus('⏹ Stopped', '');
      }
    } catch (err) {
      setStatus(`⚠ Unexpected error: ${err.message}`, 'error');
    } finally {
      running = false;
      paused  = false;
      if (currentBlockId) FlowEngine.setBlockExecuting(currentBlockId, false);
      currentBlockId = null;
      setButtonStates(false);
      setStepBadge('');
      FlowEngine.clearAllExecuting();
    }
  }

  /* ── Get next block ID from a given output port ─────────── */
  function getNext(blockId, portIndex = 0) {
    const conn = FlowEngine.getConnections().find(
      c => c.sourceId === blockId && c.sourcePort === portIndex
    );
    return conn ? conn.targetId : null;
  }

  /* ── Pause / Resume ─────────────────────────────────────── */
  function pause() {
    if (!running) return;
    paused = !paused;
    const btn = document.getElementById('btn-pause');
    if (btn) btn.textContent = paused ? '▶ Resume' : '⏸ Pause';
    if (!paused) setStatus('▶ Resumed', 'running');
  }

  /* ── Stop ───────────────────────────────────────────────── */
  function stop() {
    stopRequested = true;
    paused        = false;
    ROS.stop();
  }

  /* ── Public ─────────────────────────────────────────────── */
  return { run, pause, stop };
})();
